server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java

```diff
@@ -115,99 +126,167 @@ public void define(WebService.NewController context) {
-    String warning = getWarning(query);
-
-    SuggestionsWsResponse searchWsResponse = toResponse(componentsPerQualifiers, recentlyBrowsedKeys, favoriteKeys, warning);
-    writeProtobuf(searchWsResponse, wsRequest, wsResponse);
+      .setFavoriteKeys(favoriteKeys)
+      .setQualifiers(qualifiers);
+    if (more != null) {
+      queryBuilder.setLimit(EXTENDED_LIMIT);
+    }
+    ComponentIndexResults componentsPerQualifiers = searchInIndex(queryBuilder.build());
+    if (componentsPerQualifiers.isEmpty()) {
+      return newBuilder().build();
+    }
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Set<String> componentUuids = componentsPerQualifiers.getQualifiers()
+        .map(ComponentHitsPerQualifier::getHits)
+        .flatMap(Collection::stream)
+        .map(ComponentHit::getUuid)
+        .collect(toSet());
+      Stream<ComponentDto> componentDtoStream = dbClient.componentDao().selectByUuids(dbSession, componentUuids).stream();
+      Set<String> favoriteUuids = favorites.stream().map(ComponentDto::uuid).collect(MoreCollectors.toSet(favorites.size()));
+      SuggestionsWsResponse.Builder searchWsResponse = buildResponse(recentlyBrowsedKeys, favoriteUuids, componentsPerQualifiers, dbSession, componentDtoStream);
+      getWarning(query).ifPresent(searchWsResponse::setWarning);
+      return searchWsResponse.build();
+    }
   }
 
-  private static String getWarning(String query) {
-    List<String> tokens = ComponentTextSearchFeature.split(query).collect(Collectors.toList());
+  private static Optional<String> getWarning(String query) {
+    List<String> tokens = ComponentTextSearchFeatureRepertoire.split(query).collect(Collectors.toList());
     if (tokens.stream().anyMatch(token -> token.length() < MINIMUM_NGRAM_LENGTH)) {
-      return SHORT_INPUT_WARNING;
+      return Optional.of(SHORT_INPUT_WARNING);
     }
-    return null;
+    return Optional.empty();
   }
 
-  private ComponentIndexResults getComponentsPerQualifiers(@Nullable String more, ComponentIndexQuery.Builder queryBuilder) {
-    List<String> qualifiers;
+  private static List<String> getQualifiers(@Nullable String more) {
     if (more == null) {
-      qualifiers = stream(SuggestionCategory.values()).map(SuggestionCategory::getQualifier).collect(Collectors.toList());
-    } else {
-      qualifiers = singletonList(SuggestionCategory.getByName(more).getQualifier());
-      queryBuilder.setLimit(EXTENDED_LIMIT);
+      return stream(SuggestionCategory.values()).map(SuggestionCategory::getQualifier).collect(Collectors.toList());
     }
-    queryBuilder.setQualifiers(qualifiers);
-    return searchInIndex(queryBuilder.build());
+    return singletonList(SuggestionCategory.getByName(more).getQualifier());
+  }
+
+  private SuggestionsWsResponse.Builder buildResponse(Set<String> recentlyBrowsedKeys, Set<String> favoriteUuids, ComponentIndexResults componentsPerQualifiers,
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Split this 181 characters long line (which is greater than 180 authorized). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)